### PR TITLE
8292301: [REDO v2] C2 crash when allocating array of size too large

### PR DESCRIPTION
**Pull request by rwestrel**
On top of the redo, this fixed 2 bugs:

8288184: the problem here is that the ValidLengthTest input of an
AllocateArrayNode becomes a constant. The CatchNode would then change
types if it was reprocessed but it's not. Custom logic is needed to
enqueue the CatchNode when the ValidLengthTest input of an
AllocateArrayNode changes. The CastII out of the AllocateArrayNode
becomes top but the fallthrough path doesn't die. This happens with
igvn in the case of the bug but could also happen with ccp. I fixed
both in this patch.

8291665: the code pattern for this is 2 AllocateArrayNodes out of loop
with a shared ValidLengthTest input in a loop. When the loop is cloned
that causes Phis to be added between the AllocateArrayNodes and the
BoolNode of the ValidLengthTest inputs. Split if runs next and it
doesn't expect the Phi at the ValidLengthTest inputs. The fix here is
to clone the Bool/Cmp subgraph down on loop cloning. There's logic for
that when the use of the bool is an If for instance so I simply added
a special case to run that logic for an AllocateArrayNode use as
well. Note that the test case I added fails reliably on 11 but not
with the current jdk developement branch. AFAICT, the bug is there but
something unrelated changed and a slightly different graph is built
for the test case that prevents split if.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292301](https://bugs.openjdk.org/browse/JDK-8292301): [REDO v2] C2 crash when allocating array of size too large

### Reviewers
 * [Xin Liu](https://openjdk.org/census#xliu) (@navyxliu - Committer) ⚠️ Review applies to [9d92011a](https://git.openjdk.org/jdk/pull/10038/files/9d92011add07dbfb5ad609f63166f47e969b89ba)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [9d92011a](https://git.openjdk.org/jdk/pull/10038/files/9d92011add07dbfb5ad609f63166f47e969b89ba)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10038/head:pull/10038` \
`$ git checkout pull/10038`

Update a local copy of the PR: \
`$ git checkout pull/10038` \
`$ git pull https://git.openjdk.org/jdk pull/10038/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10038`

View PR using the GUI difftool: \
`$ git pr show -t 10038`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10038.diff">https://git.openjdk.org/jdk/pull/10038.diff</a>

</details>

### CODE REVIEW

**Comment by navyxliu** [comment]
This PR compounds 3 patches. The base is [JDK-9279219](https://github.com/openjdk/jdk/pull/6952), which has reviewed. On top of that, two others patches are bugfixes for the corner cases.

File: src/hotspot/share/opto/phaseX.cpp [hunk 1]
```diff
@@ -1843,8 +1850,9 @@ void PhaseCCP::push_phis(Unique_Node_List& worklist, const Node* use) const {
 
 // If we changed the receiver type to a call, we need to revisit the Catch node following the call. It's looking for a
 // non-NULL receiver to know when to enable the regular fall-through path in addition to the NullPtrException path.
+// Same if true if the type of a ValidLengthTest input to an AllocateArrayNode changes
 void PhaseCCP::push_catch(Unique_Node_List& worklist, const Node* use) {
-  if (use->is_Call()) {
+  if (use->is_Call() || use->is_AllocateArray()) {
```
Comment (navyxliu) [inline src/hotspot/share/opto/phaseX.cpp:1855]: hi, Roland, 
I understand your intention here, but isn't AllocateArrayNode also a CallNode? 
My understanding is that use->is_Call() is true if use is an AllocationNode.

**Review by navyxliu — COMMENTED** [review]
the two bugfixes work for me. LGTM.  
I am not a reviewer. need other reviewers to approve it.

**Comment by rwestrel** [comment]
> the two bugfixes work for me. LGTM. I am not a reviewer. need other reviewers to approve it.

Thanks for reviewing this.

**Review by navyxliu — APPROVED** [review]
LGTM.

File: src/hotspot/share/opto/phaseX.cpp [hunk 2]
```diff
@@ -1843,6 +1850,7 @@ void PhaseCCP::push_phis(Unique_Node_List& worklist, const Node* use) const {
 
 // If we changed the receiver type to a call, we need to revisit the Catch node following the call. It's looking for a
 // non-NULL receiver to know when to enable the regular fall-through path in addition to the NullPtrException path.
+// Same if true if the type of a ValidLengthTest input to an AllocateArrayNode changes
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/phaseX.cpp:1853]:
```suggestion
// Same is true if the type of a ValidLengthTest input to an AllocateArrayNode changes.
```

File: src/hotspot/share/opto/phaseX.cpp [hunk 3]
```diff
@@ -1640,6 +1640,13 @@ void PhaseIterGVN::add_users_to_worklist( Node *n ) {
         if (imem != NULL)  add_users_to_worklist0(imem);
       }
     }
+    if (use_op == Op_AllocateArray && n == use->in(AllocateNode::ValidLengthTest)) {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/phaseX.cpp:1643]: Please add a comment.

File: src/hotspot/share/opto/loopopts.cpp [hunk 4]
```diff
@@ -2406,9 +2407,10 @@ void PhaseIdealLoop::finish_clone_loop(Node_List* split_if_set, Node_List* split
   if (split_if_set) {
     while (split_if_set->size()) {
       Node *iff = split_if_set->pop();
-      if (iff->in(1)->is_Phi()) {
-        Node *b = clone_iff(iff->in(1)->as_Phi());
-        _igvn.replace_input_of(iff, 1, b);
+      uint input = iff->Opcode() == Op_AllocateArray ? AllocateNode::ValidLengthTest : 1;
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopopts.cpp:2410]:
```suggestion
      uint input = (iff->Opcode() == Op_AllocateArray) ? AllocateNode::ValidLengthTest : 1;
```

File: src/hotspot/share/opto/loopopts.cpp [hunk 5]
```diff
@@ -2037,7 +2037,8 @@ void PhaseIdealLoop::clone_loop_handle_data_uses(Node* old, Node_List &old_new,
       // in the loop to break the loop, then test is again outside of the
       // loop to determine which way the loop exited.
       // Loop predicate If node connects to Bool node through Opaque1 node.
-      if (use->is_If() || use->is_CMove() || C->is_predicate_opaq(use) || use->Opcode() == Op_Opaque4) {
+      if (use->is_If() || use->is_CMove() || C->is_predicate_opaq(use) || use->Opcode() == Op_Opaque4 ||
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopopts.cpp:2040]: Please add a comment describing the new case.

**Review by TobiHartmann — APPROVED** [review]
Looks good otherwise.

File: src/hotspot/share/opto/loopopts.cpp (hunk 4 above)
Comment (rwestrel) [reply src/hotspot/share/opto/loopopts.cpp:2422]: I left that one out as I think this pattern is common enough that it shouldn't be ambiguous.

File: src/hotspot/share/opto/phaseX.cpp (hunk 3 above)
Comment (rwestrel) [reply src/hotspot/share/opto/phaseX.cpp:1646]: Done in an updated change.

File: src/hotspot/share/opto/loopopts.cpp (hunk 5 above)
Comment (rwestrel) [reply src/hotspot/share/opto/loopopts.cpp:2045]: Done in updated change.

**Comment by rwestrel** [comment]
> Looks good otherwise.

Thanks for reviewing.

**Review by TobiHartmann — APPROVED** [review]
Thanks for adding these comments. Looks good!

**Review by navyxliu — CHANGES_REQUESTED** [review]
still LGTM.

**Review by navyxliu — APPROVED** [review]
LGTM.

**Comment by rwestrel** [comment]
> LGTM.

Thanks for re-reviewing. @TobiHartmann recommended one more review as this change caused issues and had to be backed out.

File: src/hotspot/share/opto/compile.cpp [hunk 6]
```diff
@@ -4022,6 +4025,14 @@ bool Compile::final_graph_reshaping() {
         record_method_not_compilable("malformed control flow");
         return true;            // Not all targets reachable!
       }
+    } else if (n->is_PCTable() && n->in(0) && n->in(0)->in(0) && n->in(0)->in(0)->is_Call()) {
+      CallNode* call = n->in(0)->in(0)->as_Call();
+      if (call->entry_point() == OptoRuntime::new_array_Java() ||
+          call->entry_point() == OptoRuntime::new_array_nozero_Java()) {
+        assert(call->is_CallStaticJava(), "static call expected");
+        assert(call->req() == call->jvms()->endoff() + 1, "missing extra input");
+        call->del_req(call->req()-1); // valid length test useless now
```
Comment (vnkozlov) [inline src/hotspot/share/opto/compile.cpp:4034]: How you remove graph (test) attached to ValidLengthTest input? How you know that `call` node still has this input and it is `call->req() -1`?

File: src/hotspot/share/opto/compile.cpp [hunk 7]
```diff
@@ -3998,22 +3998,25 @@ bool Compile::final_graph_reshaping() {
...
+                     call->entry_point() == OptoRuntime::new_array_nozero_Java()) {
+            // Check for illegal array length. In such case, the optimizer has
             // detected that the allocation attempt will always result in an
             // exception. There is no fall-through projection of this CatchNode .
-            Node *arg1 = call->in(TypeFunc::Parms+1);
-            if (arg1->is_Type() &&
-                arg1->as_Type()->type()->join(TypeInt::POS)->empty()) {
+            assert(call->is_CallStaticJava(), "static call expected");
+            assert(call->req() == call->jvms()->endoff() + 1, "missing extra input");
+            Node* valid_length_test = call->in(call->req()-1);
```
Comment (vnkozlov) [inline src/hotspot/share/opto/compile.cpp:4013]: How you know that call node still has this input and it is call->req() -1?

File: src/hotspot/share/opto/macro.cpp [hunk 8]
```diff
@@ -1377,6 +1378,12 @@ void PhaseMacroExpand::expand_allocate_common(
   // Copy debug information and adjust JVMState information, then replace
   // allocate node with the call
   call->copy_call_debug_info(&_igvn, alloc);
+  // For array allocations, copy the valid length check to the call node so Compile::final_graph_reshaping() can verify
+  // that the call has the expected number of CatchProj nodes (in case the allocation always fails and the fallthrough
+  // path dies).
+  if (valid_length_test != NULL) {
```
Comment (vnkozlov) [inline src/hotspot/share/opto/macro.cpp:1384]: Should we check for TOP too?

**Review by vnkozlov — COMMENTED** [review]
Should we move `check_no_dead_use()` call after `final_graph_reshaping()` call because we may have dead uses?

**Comment by vnkozlov** [comment]
I don't see link to mach5 pre-integration testing in JBS. Did we run it or we waiting final version?

File: src/hotspot/share/opto/compile.cpp (hunk 6 above)
Comment (rwestrel) [reply src/hotspot/share/opto/compile.cpp]: You're right that the subgraph at ValidLengthTest goes dead and is not properly removed. I pushed a new change to fix that.
The ValidLengthTest input from the AllocateArrayNode is always moved to the call so there's no reason it shouldn't still be there.

File: src/hotspot/share/opto/macro.cpp (hunk 8 above)
Comment (rwestrel) [reply src/hotspot/share/opto/macro.cpp:1384]: valid_length_test is NULL for AllocateNode and not NULL for AllocateArrayNode. PhaseMacroExpand::expand_allocate_common() is shared by the 2 node types and that test is used to differentiate the two. I don't think top is possible here. I also think for simplicity we want to always move the ValidLength input from the AllocateArrayNode to the call.

**Comment by rwestrel** [comment]
Thanks for reviewing this.

> Should we move `check_no_dead_use()` call after `final_graph_reshaping()` call because we may have dead uses?

Yes, but we would then hit: https://bugs.openjdk.org/browse/JDK-8211759 
That is, there are other transformations in final graph reshape that don't properly collect dead nodes so I think this should be done as a standalone change.

**Comment by TobiHartmann** [comment]
> I don't see link to mach5 pre-integration testing in JBS. Did we run it or we waiting final version?

I re-submitted testing and will report back once it passed.

File: test/hotspot/jtreg/compiler/allocation/TestFailedAllocationBadGraph.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+ */
+
+/*
+ * @test
+ * bug 8279219
+ * @summary C2 crash when allocating array of size too large
+ * @requires vm.compiler2.enabled
+ * @library /test/lib /
+ * @build sun.hotspot.WhiteBox
+ * @run driver jdk.test.lib.helpers.ClassFileInstaller sun.hotspot.WhiteBox
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/allocation/TestFailedAllocationBadGraph.java:31]: The test fails with
```
test result: Error. can't find sun.hotspot.WhiteBox in test directory or libraries
```
It should use `jdk.test.whitebox.WhiteBox`.

Comment (rwestrel) [reply test/hotspot/jtreg/compiler/allocation/TestFailedAllocationBadGraph.java]: Thanks. Fixed now.

**Comment by vnkozlov** [comment]
> Thanks for reviewing this.
> 
> > Should we move `check_no_dead_use()` call after `final_graph_reshaping()` call because we may have dead uses?
> 
> Yes, but we would then hit: https://bugs.openjdk.org/browse/JDK-8211759 That is, there are other transformations in final graph reshape that don't properly collect dead nodes so I think this should be done as a standalone change.

Okay. I agree on separate changes.

File: src/hotspot/share/opto/macro.cpp (hunk 8 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/macro.cpp:1384]: Got it.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by TobiHartmann** [comment]
I'll re-run testing and report back once it passed.

**Comment by TobiHartmann** [comment]
All tests passed.

**Comment by rwestrel** [comment]
thanks for the review @vnkozlov and testing @TobiHartmann

**Pull request closed without merging**